viewing_party/party.py

```diff
@@ -1,23 +1,151 @@
+            user_data['watched'].append(movie)
+            # return user_data
+        
+    return user_data  
+# # -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
+def get_watched_avg_rating(user_data):
+    rating_list = []
+    for i in range(len(user_data["watched"])):
+        rating_list.append(user_data["watched"][i]["rating"])
+    if rating_list:
+        average = sum(rating_list)/len(rating_list)
+    else:
+        average = 0.0
+    return average 
 
 
+def get_most_watched_genre(user_data):
+    genre_list = []
+    for i in range(len(user_data["watched"])):
+        genre_list.append(user_data["watched"][i]["genre"])
+    if genre_list: 
+        for genre in genre_list:
+            popular_genre = max(set(genre_list), key = genre_list.count)
+    else: 
+        popular_genre = None 
+    return popular_genre
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
+#----- Extra helper function to get friends_watched_list ----
+def get_friends_watched_list(user_data):
+    friends_watched_list = []
+    for friend in user_data["friends"]:
+            for movie in friend["watched"]:
+                friends_watched_list.append(movie["title"])
+    return friends_watched_list
 
-        
+
+def get_unique_watched(user_data): 
+    friends_watched_list = get_friends_watched_list(user_data)
+    unique_watched_list = []
+    
+    for movie in user_data["watched"]:
+        if movie["title"] not in friends_watched_list:  
+            unique_watched_list.append(movie)
+    
+    return unique_watched_list
+
+
+def get_friends_unique_watched(user_data):
+    friends_unique_watched_list = []
+    user_list = []
+    friends_unique_movies = []
+    friends = user_data["friends"]
+    user_watched = user_data["watched"]
+
+    for movie in user_watched:
+        user_list.append(movie["title"])
+    
+    for friend in friends:
+        for movie in friend["watched"]:
+            if movie["title"] not in user_list:
+                friends_unique_watched_list.append(movie)
+    
+    for i in range(len(friends_unique_watched_list)):
+        if friends_unique_watched_list[i] not in friends_unique_watched_list[i + 1:]:
+            friends_unique_movies.append(friends_unique_watched_list[i])
+
+    return friends_unique_movies
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
-
+def get_available_recs(user_data):
+    friend_recs = []
+    final_recs = []
+    user_watched_list = [movies for movies in user_data["watched"]]
+    
+    for movie in user_data["friends"]:
+        for title in movie["watched"]:
+            if title not in user_watched_list:
+                if title not in friend_recs:
+                    friend_recs.append(title)

```

Comment:
careful with your variable names here! we want them to describe what is inside the data structure for other coders to understand, as well as you in a few months when you come back to an old project.

On line 107, are we looking at a `movie` in `user_data["friends"]` or a single friend object?. Let's call this `for friend in user_data["friends"]`.

Now what about line 108? Is it a title or is it a movie?